frontends/p4/sideEffects.cpp

```diff
@@ -391,39 +432,282 @@ const IR::Node* DoSimplifyExpressions::preorder(IR::MethodCallExpression* mce) {
         }
     }
 
-    if (anyOut) {
-        // Check aliasing between all pairs of arguments where at
-        // least one of them is out or inout.
-        for (auto p1 : *mi->substitution.getParametersInArgumentOrder()) {
-            auto arg1 = mi->substitution.lookup(p1);
-            if (hasSideEffects.count(arg1->expression))
+    // Check aliasing between all pairs of arguments where at
+    // least one of them is out or inout.
+    for (auto p1 : *mi->substitution.getParametersInArgumentOrder()) {
+        if (p1->direction == IR::Direction::None)
+           continue;
+        auto arg1 = mi->substitution.lookup(p1);
+
+        for (auto p2 : *mi->substitution.getParametersInArgumentOrder()) {
+            if (p2->direction == IR::Direction::None)
                 continue;
-            for (auto p2 : *mi->substitution.getParametersInArgumentOrder()) {
-                LOG3("p1=" << dbp(p1) << " p2=" << dbp(p2));
-                if (p2 == p1)
-                    break;
-                if (!p1->hasOut() && !p2->hasOut())
+
+            LOG3("p1=" << dbp(p1) << " p2=" << dbp(p2));
+            if (p2 == p1)
+                break;
+
+            auto arg2 = mi->substitution.lookup(p2);
+
+            if (auto m = arg1->expression->to<IR::Member>()) {
+                auto u = arg1->expression->to<IR::Operation_Unary>();
+                if  (u->expr->is<IR::ArrayIndex>()) {
+                    auto b = u->expr->to<IR::Operation_Binary>();
+                    // we handle this case T f(in T x, inout T y){...}; T g(inout T z){...}
+                    // f(g(val), s[val].x)
```

Comment:
this comment does not seem to match the code. The code checks if arg1 is that shape a[].m

Reply:
It may not be so obvious (it's not intuitive) but nested for-loop visit arguments in that order.
Each argument is processed with all previous arguments.

For example:
Let f(a, b, c, d, e, f)
arg1: b; arg2: a;
arg1: c; arg2: a;
arg1: c; arg2: b;
arg1: d; arg2: a;
arg1: d; arg2: b;
arg1: d; arg2: c;
arg1: e; arg2: a;
arg1: e; arg2: b;
arg1: e; arg2: c;
arg1: e; arg2: d;

That is why my comment matched the code.
(for f(g(val), s[val].x)  
arg1: s[val].x; arg2: g(val); )